docs/architecture/adr-082-socket-protocol.md

```diff
@@ -0,0 +1,167 @@
+  error condition is terminal not only to the request, but to the entire ABCI
+  connection.
+
+  The intent historically seems to have been that all ABCI errors are protocol
+  fatal. In practice, however, this greatly complicates debugging a faulty
+  node, since the only way to respond to errors it to panic the node which
+  loses valuable context that could have been logged.
+
+- There are subtle concurrency management dependencies between the client and
+  the server that are not clearly documented anywhere, and it is very easy for
+  small changes in both the client and the server to lead to tricky deadlocks,
+  panics, race conditions, and slowdowns. As a recent example of this, see
+  https://github.com/tendermint/tendermint/pull/8581.
+
+These limitations are fixable, but one important question is whether it is
+worthwhile to fix them.  The stability and language coverage of gRPC have
+improved a lot, and today it is probably simpler to set up and maintain an
+application using gRPC transport than to reimplement the Tendermint socket
+protocol.
+
+Moreover, gRPC addresses all the above issues out-of-the-box, and requires
+(much) less custom code for both the server (i.e., the application) and the
+client. The project is well-funded and widely-used, which makes it a safe bet
+for a dependency.
+
+## Decision
+
+There is a set of related alternatives to consider:
+
+- Question 1: Designate a single IPC standard for out-of-process applications?
+
+  Claim: We should converge on one (and only one) IPC option for out-of-process
+  applications. We should choose an option that, after a suitable period of
+  deprecation, can address most or all the highest-impact uses of Tendermint.
+  Maintaining multiple options increases the surface area for bugs and
+  vulnerabilities, and we should not have multiple options for basic interfaces
+  without a clear and well-documented reason.
+
+- Question 2a: Standardize on gRPC and deprecate/remove the socket protocol?
+
+  Claim: Maintaining and improving a custom RPC protocol is a substantial
+  project and not directly relevant to the requirements of consensus. We would
+  be better served by depending on a well-maintained open-source library like
+  gRPC.
+
+- Question 2b: Improve the socket protocol and deprecate/remove gRPC?
+
+  Claim: If we find meaningful advantages to maintaining our own custom RPC
+  protocol in Tendermint, we should treat it as a first-class project within
+  the core and invest in making it good enough that we do not require other
+  options.
+
+## Detailed Design
+
+1. If we choose to standardize on gRPC, the main work in in Tendermint core
+   will be removing and cleaning up the code for the socket client and server.
+
+   Besides the code cleanup, we will also need to clearly document a
+   deprecation schedule, and invest time in making the migration easier for
+   applications currently using the socket protocol.
```

Comment:
Do you think it's worth adding a migrations strategies section to this document? I'm not sure I have a clear picture of what a migration may look like, but I think understanding how much we intended to commit to helping migrations may make the cost/benefit trade-off more clear.

Reply:
Worthwhile yes, but I'm not sure I have a clear picture yet either. In theory as long as the ABCI protocol itself doesn't change, this mainly involves updating internal plumbing, but it would presumably need to be a coordinated upgrade across the whole network. (I don't know if there are networks with the same node logic implemented in two different languages—it seems unlikely but I suppose is theoretically possible) 

So I _believe_ the main work is wiring up that plumbing and then testing the bejeezus out of it to make sure nothing semantic changed. But I have observed that applications tend to be pathological about ignoring basic abstraction boundaries so I'd really want input from real network operators.

Reply:
I added some text to the document to highlight this question.

Reply:
This is somewhat orthogonal but I believe the SDK gives the option for a node to run in-process and out-of-process.